src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,224 @@
+    // /// used for `accelerationStructureCaptureReplay`
+    // device_address: VkDeviceAddress,
+}
+
+bitflags! {
+    /// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkBuildAccelerationStructureFlagBitsKHR.html
+    pub struct AccelerationStructureFlags: u32 {
+        /// indicates that the specified acceleration structure can be updated with update of VK_TRUE in vkCmdBuildAccelerationStructuresKHR or vkCmdBuildAccelerationStructureNV .
+        const ALLOW_UPDATE = 0x1;
+        ///  indicates that the specified acceleration structure can act as the source for a copy acceleration structure command with mode of VK_COPY_ACCELERATION_STRUCTURE_MODE_COMPACT_KHR to produce a compacted acceleration structure.
+        const ALLOW_COMPACTION = 0x2;
+        /// indicates that the given acceleration structure build should prioritize trace performance over build time.
+        const PREFER_FAST_TRACE = 0x4;
+        ///  indicates that the given acceleration structure build should prioritize build time over trace performance.
+        const PREFER_FAST_BUILD = 0x8;
+        /// indicates that this acceleration structure should minimize the size of the scratch memory and the final result build, potentially at the expense of build time or trace performance.
+        const LOW_MEMORY = 0x10;
+    }
+}
+
+#[derive(Debug)]
+pub enum AccelerationStructureCreateMode {
+    /// specifies that the destination acceleration structure will be built using the specified geometries.
+    Build,
+    /// specifies that the destination acceleration structure will be built using data in a source acceleration structure, updated by the specified geometries.
+    /// Note there's constraints on update: https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#acceleration-structure-update-constraints
+    Update,
+}
+
+/// A description of the geometry data needed to populate an acceleration structure.
+///
+/// TODO: there's something here that smells w/ what fields are needed to get the required build size vs what fields are needed to actually build. Also, the top/bottom levels having different requirements on which fields are valid.
+// TODO: I don't like that this is reused for the "get build sizes" and "do actual build" with complex rules on what fields are ignored when. Perhaps we could use DX12's model for `D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_DESC` and `D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS` and move at least the src/dst fields onto the actual "build acceleration structure" command.?
+#[derive(Debug)]
+pub struct AccelerationStructureGeometryDesc<'a, B: Backend> {
+    pub flags: AccelerationStructureFlags,
+    /// The type of acceleration structure to build.
+    pub ty: AccelerationStructureType,
+    pub mode: AccelerationStructureCreateMode,
+
+    pub src: Option<B::AccelerationStructure>,
+    pub dst: Option<B::AccelerationStructure>,
+
+    // TODO: We could enforce the following the type system?
```

Comment:
Seems worth a try to see how this would look.

A small tale from a similar issue with had with descriptor updates. Vulkan requires all the descriptors within a "write" to be of the same type (and other parameters). The backends didn't end up taking advantage of this though, and `wgpu-core` as the main user didn't enjoy this either. So the restriction only put unnecessary friction between `wgpu-core` and the backends, making the API not do its function best for this path.

In the end, we just relaxed the rules in gfx-hal itself and allowed different types of descriptors to be passed in. This required almost no extra logic in Vulkan backend, and it resolved a concern on whether we should be type-enforcing this.

Reply:
I totally get that. I believe this is a case where all the backends will have this same requirement and it could be impractical to implement a relaxed version (we'd potentially need to partition an acceleration structure into multiple based on geometry type?). Note that nothing prevents vulkan or dx from being implemented as it's listed currently.

I suspect in this case we may be able to do it without much (any?) "clever" type system trickery. For reference, this is how DX12 enforces it in a C API: [D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS ](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_build_raytracing_acceleration_structure_inputs), basically moving the union tag outside of the list rather than on each element.

At the very least, without safety at the type-level, I want to make sure we get docs onto this in gfx to make sure users know about this restriction.